File: server/migrations/migrations.go
package migrations

import (
	"context"
	"crypto/sha256"
	"embed"
	"encoding/hex"
	"errors"
	"fmt"
	"io/fs"
	"log/slog"
	"regexp"
	"sort"
	"strconv"
	"time"

	"github.com/jackc/pgx/v5"
	"github.com/jackc/pgx/v5/pgxpool"
)

const advisoryLockKey int64 = 0x42657272794d6967

var migrationName = regexp.MustCompile(`^([0-9]{3})_[a-z0-9_]+\.up\.sql$`)

//go:embed *.up.sql
var files embed.FS

// Migration is one immutable, forward-only database change.
type Migration struct {
	Version  int
	Name     string
	SQL      string
	Checksum string
}

// List returns the validated embedded migrations in ascending version order.
func List() ([]Migration, error) {
	return loadFS(files)
}

func loadFS(source fs.FS) ([]Migration, error) {
	names, err := fs.Glob(source, "*.up.sql")
	if err != nil {
		return nil, fmt.Errorf("list migrations: %w", err)
	}
	sort.Strings(names)

	result := make([]Migration, 0, len(names))
	versions := make(map[int]string, len(names))
	for _, name := range names {
		match := migrationName.FindStringSubmatch(name)
		if match == nil {
			return nil, fmt.Errorf("invalid migration filename %q", name)
		}
		version, err := strconv.Atoi(match[1])
		if err != nil {
			return nil, fmt.Errorf("parse migration version %q: %w", name, err)
		}
		if previous, exists := versions[version]; exists {
			return nil, fmt.Errorf(
				"duplicate migration version %03d in %q and %q",
				version,
				previous,
				name,
			)
		}
		versions[version] = name

		body, err := fs.ReadFile(source, name)
		if err != nil {
			return nil, fmt.Errorf("read migration %q: %w", name, err)
		}
		sum := sha256.Sum256(body)
		result = append(result, Migration{
			Version:  version,
			Name:     name,
			SQL:      string(body),
			Checksum: hex.EncodeToString(sum[:]),
		})
	}
	return result, nil
}

// Apply runs all pending migrations while holding a session advisory lock.
func Apply(ctx context.Context, pool *pgxpool.Pool, logger *slog.Logger) error {
	if pool == nil {
		return errors.New("apply migrations: database pool is nil")
	}
	if logger == nil {
		logger = slog.Default()
	}
	all, err := List()
	if err != nil {
		return err
	}

	conn, err := pool.Acquire(ctx)
	if err != nil {
		return fmt.Errorf("acquire migration connection: %w", err)
	}
	defer conn.Release()

	if _, err := conn.Exec(ctx, "SELECT pg_advisory_lock($1)", advisoryLockKey); err != nil {
		return fmt.Errorf("acquire migration lock: %w", err)
	}
	defer func() {
		unlockCtx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
		defer cancel()
		if _, unlockErr := conn.Exec(
			unlockCtx,
			"SELECT pg_advisory_unlock($1)",
			advisoryLockKey,
		); unlockErr != nil {
			logger.Error("release migration lock", "error", unlockErr)
		}
	}()

	if _, err := conn.Exec(ctx, `
		CREATE TABLE IF NOT EXISTS berry_schema_migrations (
			version integer PRIMARY KEY,
			name text NOT NULL UNIQUE,
			checksum char(64) NOT NULL,
			applied_at timestamptz NOT NULL DEFAULT now(),
			duration_ms bigint NOT NULL DEFAULT 0 CHECK (duration_ms >= 0)
		)
	`); err != nil {
		return fmt.Errorf("initialize migration ledger: %w", err)
	}

	applied, err := readApplied(ctx, conn)
	if err != nil {
		return err
	}
	expected := make(map[int]Migration, len(all))
	for _, migration := range all {
		expected[migration.Version] = migration
	}
	for version, record := range applied {
		migration, ok := expected[version]
		if !ok {
			return fmt.Errorf(
				"migration ledger contains version %03d (%s) missing from this binary",
				version,
				record.name,
			)
		}
		if record.name != migration.Name {
			return fmt.Errorf(
				"migration %03d name drift: ledger has %q, binary has %q",
				version,
				record.name,
				migration.Name,
			)
		}
		if record.checksum != migration.Checksum {
			return fmt.Errorf("migration %q checksum drift", migration.Name)
		}
	}

	for _, migration := range all {
		if _, ok := applied[migration.Version]; ok {
			continue
		}
		if err := applyOne(ctx, conn, migration, logger); err != nil {
			return err
		}
	}
	return nil
}

type appliedMigration struct {
	name     string
	checksum string
}

type migrationConnection interface {
	Begin(context.Context) (pgx.Tx, error)
	Query(context.Context, string, ...any) (pgx.Rows, error)
}

func readApplied(
	ctx context.Context,
	conn migrationConnection,
) (map[int]appliedMigration, error) {
	rows, err := conn.Query(
		ctx,
		"SELECT version, name, checksum FROM berry_schema_migrations ORDER BY version",
	)
	if err != nil {
		return nil, fmt.Errorf("read migration ledger: %w", err)
	}
	defer rows.Close()

	result := make(map[int]appliedMigration)
	for rows.Next() {
		var version int
		var record appliedMigration
		if err := rows.Scan(&version, &record.name, &record.checksum); err != nil {
			return nil, fmt.Errorf("scan migration ledger: %w", err)
		}
		result[version] = record
	}
	if err := rows.Err(); err != nil {
		return nil, fmt.Errorf("iterate migration ledger: %w", err)
	}
	return result, nil
}

func applyOne(
	ctx context.Context,
	conn migrationConnection,
	migration Migration,
	logger *slog.Logger,
) error {
	started := time.Now()
	tx, err := conn.Begin(ctx)
	if err != nil {
		return fmt.Errorf("begin migration %q: %w", migration.Name, err)
	}
	defer func() {
		_ = tx.Rollback(context.Background())
	}()

	if _, err := tx.Exec(ctx, migration.SQL); err != nil {
		return fmt.Errorf("execute migration %q: %w", migration.Name, err)
	}
	duration := time.Since(started)
	if _, err := tx.Exec(
		ctx,
		`INSERT INTO berry_schema_migrations
			(version, name, checksum, duration_ms)
		 VALUES ($1, $2, $3, $4)`,
		migration.Version,
		migration.Name,
		migration.Checksum,
		duration.Milliseconds(),
	); err != nil {
		return fmt.Errorf("record migration %q: %w", migration.Name, err)
	}
	if err := tx.Commit(ctx); err != nil {
		return fmt.Errorf("commit migration %q: %w", migration.Name, err)
	}
	logger.Info(
		"applied database migration",
		"name",
		migration.Name,
		"duration",
		duration,
	)
	return nil
}

File: server-ts/src/migrate/migrations.ts
import { createHash } from 'node:crypto';
import { readFile, readdir } from 'node:fs/promises';
import path from 'node:path';
import { fileURLToPath } from 'node:url';
import type { Sql } from '../db/pool.ts';

/**
 * Forward-only schema migration.
 *
 * The ledger table, its checksums and the advisory lock key are the ones Berry
 * has always used, so a database migrated before this runner existed is
 * already current: it reads the same `berry_schema_migrations` rows and finds
 * nothing to do. That compatibility is the point — changing the runner must
 * not ask anyone to reset a database.
 */

/** `BerryMig` as a big-endian int64 — the key Berry has always locked on. */
const ADVISORY_LOCK_KEY = '4784356015137974631';

const MIGRATION_NAME = /^([0-9]{3})_[a-z0-9_]+\.up\.sql$/;

/** `server-ts/migrations`, resolved from this module rather than the cwd. */
const DIRECTORY = path.join(path.dirname(fileURLToPath(import.meta.url)), '..', '..', 'migrations');

/** One immutable, forward-only database change. */
export interface Migration {
   version: number;
   name: string;
   sql: string;
   /** SHA-256 of the file's bytes, hex — what the ledger stores. */
   checksum: string;
}

export interface Logger {
   info(message: string, fields?: Record<string, unknown>): void;
   error(message: string, fields?: Record<string, unknown>): void;
}

/** The validated migrations in ascending version order. */
export async function list(directory: string = DIRECTORY): Promise<Migration[]> {
   const names = (await readdir(directory)).filter((name) => name.endsWith('.up.sql')).sort();

   const migrations: Migration[] = [];
   const versions = new Map<number, string>();
   for (const name of names) {
      const match = MIGRATION_NAME.exec(name);
      if (!match) throw new Error(`invalid migration filename ${JSON.stringify(name)}`);

      const version = Number(match[1]);
      const previous = versions.get(version);
      if (previous !== undefined) {
         throw new Error(
            `duplicate migration version ${match[1]} in ${JSON.stringify(previous)} and ${JSON.stringify(name)}`
         );
      }
      versions.set(version, name);

      // Hashed as bytes, not as a decoded string: the checksum has to match
      // the one already recorded for the file, and a re-encode could differ.
      const body = await readFile(path.join(directory, name));
      migrations.push({
         version,
         name,
         sql: body.toString('utf8'),
         checksum: createHash('sha256').update(body).digest('hex'),
      });
   }
   return migrations;
}

interface AppliedMigration {
   name: string;
   checksum: string;
}

/**
 * Applies every pending migration while holding a session advisory lock.
 *
 * The lock is what makes two processes starting at once safe — the second
 * blocks rather than racing the first through the same CREATE TABLE. It is
 * taken on a reserved connection because a pool would hand the unlock to a
 * different one, which PostgreSQL treats as not holding the lock at all.
 */
export async function apply(sql: Sql, logger: Logger): Promise<void> {
   const all = await list();

   const conn = await sql.reserve();
   try {
      await conn.unsafe(`SELECT pg_advisory_lock(${ADVISORY_LOCK_KEY})`);
      try {
         await conn.unsafe(`
            CREATE TABLE IF NOT EXISTS berry_schema_migrations (
               version integer PRIMARY KEY,
               name text NOT NULL UNIQUE,
               checksum char(64) NOT NULL,
               applied_at timestamptz NOT NULL DEFAULT now(),
               duration_ms bigint NOT NULL DEFAULT 0 CHECK (duration_ms >= 0)
            )
         `);

         const applied = await readApplied(conn);
         const expected = new Map(all.map((migration) => [migration.version, migration]));

         // Drift is fatal rather than repaired. A ledger row this runner
         // cannot account for means the database was migrated by something
         // else, and applying more on top of it would compound the divergence.
         for (const [version, record] of applied) {
            const padded = String(version).padStart(3, '0');
            const migration = expected.get(version);
            if (!migration) {
               throw new Error(
                  `migration ledger contains version ${padded} (${record.name}) missing from this build`
               );
            }
            if (record.name !== migration.name) {
               throw new Error(
                  `migration ${padded} name drift: ledger has ${JSON.stringify(record.name)}, build has ${JSON.stringify(migration.name)}`
               );
            }
            if (record.checksum !== migration.checksum) {
               throw new Error(`migration ${JSON.stringify(migration.name)} checksum drift`);
            }
         }

         for (const migration of all) {
            if (applied.has(migration.version)) continue;
            await applyOne(conn, migration, logger);
         }
      } finally {
         // Best effort: the lock dies with the session anyway, so a failure
         // here is worth reporting but must not mask the original error.
         await conn
            .unsafe(`SELECT pg_advisory_unlock(${ADVISORY_LOCK_KEY})`)
            .catch((error: unknown) => logger.error('release migration lock', { error: String(error) }));
      }
   } finally {
      await conn.release();
   }
}

async function readApplied(conn: Sql): Promise<Map<number, AppliedMigration>> {
   const rows = await conn<{ version: number; name: string; checksum: string }[]>`
      SELECT version, name, checksum FROM berry_schema_migrations ORDER BY version
   `;
   return new Map(rows.map((row) => [row.version, { name: row.name, checksum: row.checksum }]));
}

/**
 * Runs one migration and records it, in a single transaction.
 *
 * The transaction is driven with explicit statements rather than `sql.begin`,
 * because a connection from `reserve()` does not have it: postgres.js builds
 * the reserved handle from the bare query tag, so `.begin` is absent at
 * runtime even though the typings declare `ReservedSql extends Sql`. Calling it
 * fails with `conn.begin is not a function` on the first migration.
 *
 * It has to be this connection either way — the advisory lock is session-scoped
 * and held here, so migrating on a pooled connection would apply the schema
 * outside the lock that is supposed to be protecting it.
 */
async function applyOne(conn: Sql, migration: Migration, logger: Logger): Promise<void> {
   const started = Date.now();
   await conn.unsafe('BEGIN');
   try {
      // `.simple()` is required: a migration is many statements in one file,
      // and the extended protocol accepts only one per round trip.
      await conn.unsafe(migration.sql).simple();
      const durationMs = Date.now() - started;
      await conn`
         INSERT INTO berry_schema_migrations (version, name, checksum, duration_ms)
         VALUES (${migration.version}, ${migration.name}, ${migration.checksum}, ${durationMs})
      `;
      await conn.unsafe('COMMIT');
   } catch (error) {
      // The connection is in a failed transaction and would reject everything
      // that followed, including the unlock, so this rollback matters even
      // though the process is about to exit.
      await conn.unsafe('ROLLBACK').catch(() => {});
      throw error;
   }
   logger.info('applied database migration', {
      name: migration.name,
      durationMs: Date.now() - started,
   });
}
